inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/dataprovider/generic/DataProviderGenerator.java

```diff
@@ -0,0 +1,170 @@
+     * The provider is either dynamically compiled and injected or a cached provider is used.
+     *
+     * @param providerConfig       the configuration of the generic data provider to use
+     * @param classToUseProviderOn the context in which the provider will be actively. The provider will be injected into the classloader of this class.
+     * @return the generated provider
+     * @throws ExecutionException
+     */
+    @SuppressWarnings("unchecked")
+    public InjectedClass<? extends IGenericDataProvider> getOrGenerateDataProvider(ResolvedGenericDataProviderConfig providerConfig, Class<?> classToUseProviderOn) {
+        ClassLoader loader = Optional.ofNullable(classToUseProviderOn.getClassLoader()).orElse(BOOTSTRAP_LOADER);
+        providersCache.cleanUp();
+        Cache<ResolvedGenericDataProviderConfig, InjectedClass<? extends IGenericDataProvider>> clCache;
+        try {
+            clCache = providersCache.get(loader);
+            clCache.cleanUp();
+            try {
+                return clCache.get(providerConfig, () ->
+                        (InjectedClass<? extends IGenericDataProvider>)
+                                classInjector.inject(GENERIC_PROVIDER_STRUCTURAL_ID, classToUseProviderOn, (className) ->
+                                        buildGenericDataProviderByteCode(providerConfig, loader, className)
+                                ));
+            } catch (ExecutionException | ExecutionError e) {
+                log.error("Error creating data provider {} in context of class {}! Using a No-Operation data provider instead!",
+                        providerConfig.getName(), classToUseProviderOn.getName(), e);
+                return clCache.get(providerConfig, () -> new InjectedClass<IGenericDataProvider>(GenericDataProviderTemplate.class));
+            }
+        } catch (ExecutionException e) {
+            //never happens
+            throw new RuntimeException(e);
+        }
+    }
+
+    private byte[] buildGenericDataProviderByteCode(ResolvedGenericDataProviderConfig providerConfig, ClassLoader loader, String className) throws NotFoundException, CannotCompileException, IOException {
+        ClassPool cp = new ClassPool();
+        cp.insertClassPath(new ClassClassPath(GenericDataProviderTemplate.class));
+        if (loader != BOOTSTRAP_LOADER) {
+            cp.insertClassPath(new LoaderClassPath(loader));
+        }
+
+        CtClass provider = cp.get(GenericDataProviderTemplate.class.getName());
+        provider.setName(className);
+
+        for (String packageName : providerConfig.getImportedPackages()) {
+            cp.importPackage(packageName);
+        }
+
+        CtMethod method = provider.getDeclaredMethod("executeImpl");
+        method.setBody(buildProviderMethod(providerConfig));
+
+        return provider.toBytecode();
+    }
+
+    private String buildProviderMethod(ResolvedGenericDataProviderConfig providerConfig) {
```

Comment:
we need a comment here (or javadoc) explaining how does the source of the full method look (if all options are enabled).. this is not really readable :cry: